### PR TITLE
feat: Notification Email Templates

### PR DESCRIPTION
**Pull request by TimVanOnckelen**
This PR introduces a feature to allow system users to customize the text of email notifications by setting up custom templates in the system settings. Users can tailor their email content with dynamic tags, and HTML is supported for formatting. This update aims to improve flexibility and personalization in user communications.

### Key Changes

- Added a new tab in system settings for custom email templates
- Enabled HTML support and dynamic tag usage within templates
- Preview on email templates in the backend
- Updated documentation to reflect these changes
- Added localization support for English (EN) and Dutch (NL)

This enhancement enables administrators to align notifications with specific branding or informational needs more effectively.
user-notifications-templates

Preview of email template

### CODE REVIEW

**Comment by alextran1502** [comment]
Without the HTML viewer for styling and such, it is very hard to check for the modification, right?

**Comment by TimVanOnckelen** [comment]
> Without the HTML viewer for styling and such, it is very hard to check for the modification, right?

It may indeed be handy to have a preview to check the modifications. The implementation right now is bare bones. A basic optimization would be to either add a wysiwyg editor, or add the ability to send a test mail?

**Comment by TimVanOnckelen** [comment]
I updated the PR with the feature to preview email templates from within the settings. It may be usefull to add a wysiwyg editor in favor of the textarea, so users do not need to write HTML. Any suggestions on what editor to use?

File: web/src/lib/components/admin-page/settings/template-settings/template-settings.svelte [hunk 1]
```diff
@@ -0,0 +1,132 @@
+<script lang="ts">
```
Comment (alextran1502) [inline web/src/lib/components/admin-page/settings/template-settings/template-settings.svelte:1]: Since we are on Svelte 5, can you use the Rune syntax? https://svelte.dev/docs/svelte/what-are-runes

**Comment by TimVanOnckelen** [comment]
The PR is updated using the Svelte 5 syntax.

File: web/src/lib/components/shared-components/settings/setting-textarea-wysiwyg.svelte [hunk 2]
```diff
@@ -0,0 +1,54 @@
+<script lang="ts">
```
Comment (alextran1502) [inline web/src/lib/components/shared-components/settings/setting-textarea-wysiwyg.svelte:1]: Can you help updating this file with Svelte 5 syntax as well?

Comment (TimVanOnckelen) [reply web/src/lib/components/shared-components/settings/setting-textarea-wysiwyg.svelte:1]: Of course, updated the file

**Comment by TimVanOnckelen** [comment]
@alextran1502 , i updated the last file to Svelte 5. Let me know if i need to do some more work on this pr.

**Comment by TimVanOnckelen** [comment]
There was a strange lint error on the translations, but after readding the keys a couple of times, the linter tests successfully.

**Comment by TimVanOnckelen** [comment]
I will temporary set this PR back to draft until the svelte 5 migration is done.

File: docs/docs/administration/email-notification.mdx [hunk 3]
```diff
@@ -19,3 +19,10 @@ You can use [this guide](/docs/guides/smtp-gmail) to use Gmail's SMTP server.
 Users can manage their email notification settings from their account settings page on the web. They can choose to turn email notifications on or off for the following events:
 
 <img src={require('./img/user-notifications-settings.png').default} width="80%" title="User notification settings" />
+
+## Notification templates
+
+You can override the default notification text with custom templates. You can uses tags to show dynamic tags in your templates.
```
Comment (danieldietzler) [inline docs/docs/administration/email-notification.mdx:25]:
```suggestion
You can override the default notification text with custom templates. You can use tags to show dynamic tags in your templates.
```
This is kind of a weird sentence. "You can use tags to show dynamic tags[...]"?

File: docs/docs/administration/email-notification.mdx [hunk 4]
```diff
@@ -19,3 +19,10 @@ You can use [this guide](/docs/guides/smtp-gmail) to use Gmail's SMTP server.
...
+Templates support HTML tags like P, BR, H1, H2, ...
```
Comment (danieldietzler) [inline docs/docs/administration/email-notification.mdx:26]: I don't think we need to enumerate those here

File: i18n/en.json [hunk 5]
```diff
@@ -1311,4 +1321,4 @@
   "yes": "Yes",
   "you_dont_have_any_shared_links": "You don't have any shared links",
   "zoom_image": "Zoom Image"
-}
+}
```
Comment (danieldietzler) [inline i18n/en.json:1324]: Please add the newline back to this file

File: i18n/en.json [hunk 6]
```diff
@@ -247,6 +247,16 @@
...
     "tag_cleanup_job": "Tag cleanup",
+    "template_email_preview": "Preview",
+    "template_email_settings": "Email",
+    "template_email_settings_description": "Manage custom email notification templates",
+    "template_email_welcome": "Welcome email template",
+    "template_email_invite_album": "Invite Album Template",
+    "template_email_update_album": "Update Album Template",
+    "template_settings": "Notification Templates",
+    "template_settings_description": "Manage custom templates for notifications.",
+    "template_email_if_empty": "If template is empty, the default email will be used.",
```
Comment (danieldietzler) [inline i18n/en.json:258]:
```suggestion
    "template_email_if_empty": "If the template is empty, the default email will be used.",
```

File: i18n/en.json [hunk 7]
```diff
@@ -247,6 +247,16 @@
...
+    "template_email_available_tags": "You can use following variables in your template: {tags}",
```
Comment (danieldietzler) [inline i18n/en.json:259]:
```suggestion
    "template_email_available_tags": "You can use the following variables in your template: {tags}",
```

File: server/src/dtos/notification.dto.ts [hunk 8]
```diff
@@ -1,3 +1,13 @@
...
 export class TestEmailResponseDto {
   messageId!: string;
 }
+export class TemplateResponseDto {
+  name!: string;
+  html!: string;
+}
+export class TemplateDto {
+  @IsString()
+  tempTemplate!: string;
```
Comment (danieldietzler) [inline server/src/dtos/notification.dto.ts:12]: Why is it called `tempTemplate`?

File: server/src/services/notification.service.ts [hunk 9]
```diff
@@ -175,14 +174,78 @@ export class NotificationService extends BaseService {
     return { messageId };
   }
 
+  async getTemplate(name: string, tempTemplate: string) {
+    const { server, templates } = await this.getConfig({ withCache: false });
+    const { port } = this.configRepository.getEnv();
+
+    let templateResponse = '';
+
+    switch (name) {
```
Comment (danieldietzler) [inline server/src/services/notification.service.ts:183]: That name should probably be an enum

File: server/src/utils/replace-template-tags.ts
Comment (danieldietzler) [inline server/src/utils/replace-template-tags.ts:1]: I feel like we shouldn't write this ourselves but use an existing, proper library that also takes care of properly escaping malicious code (although you may argue that an admin is trusted, not sure)

File: server/src/services/base.service.ts [hunk 10]
```diff
@@ -112,6 +112,7 @@ export class BaseService {
       configRepo: this.configRepository,
       metadataRepo: this.systemMetadataRepository,
       logger: this.logger,
+      notifications: this.notificationRepository,
```
Comment (danieldietzler) [inline server/src/services/base.service.ts:115]: Why did you add this?

File: server/src/emails/templates/album-invite.ts
Comment (danieldietzler) [inline server/src/emails/templates/album-invite.ts:1]: What are these used for?

File: web/src/lib/components/shared-components/settings/setting-textarea-wysiwyg.svelte
Comment (danieldietzler) [inline web/src/lib/components/shared-components/settings/setting-textarea-wysiwyg.svelte:1]: What about this is a WYSIWYG text area and not just any normal text area?

Comment (TimVanOnckelen) [reply web/src/lib/components/shared-components/settings/setting-textarea-wysiwyg.svelte:1]: It is indeed a normal text area, but this component functions as a placeholder. There is no wysiwyg editor introduced yet.

File: server/src/emails/templates/album-invite.ts
Comment (TimVanOnckelen) [reply server/src/emails/templates/album-invite.ts:1]: These can be used in the future to initialize default templates. But maybe that's too early at this stage?

File: server/src/utils/replace-template-tags.ts
Comment (TimVanOnckelen) [reply server/src/utils/replace-template-tags.ts:1]: Could indeed be a solution to use an external library. I had the idea to use as little overhead as possible, but in the long run it is probably better to do this properly right away. Any suggestions on what library to use?

Comment (danieldietzler) [reply server/src/utils/replace-template-tags.ts:1]: I was about to suggest the library we use for the storage template, but turns out we do it manually there as well 😅. @jrasm91 thoughts?

File: server/src/emails/templates/album-invite.ts
Comment (danieldietzler) [reply server/src/emails/templates/album-invite.ts:1]: Yeah I wouldn't put stuff in this PR that's not used at all. Especially if it's literally just an empty file

File: web/src/lib/components/shared-components/settings/setting-textarea-wysiwyg.svelte
Comment (danieldietzler) [reply web/src/lib/components/shared-components/settings/setting-textarea-wysiwyg.svelte:1]: IMO just use the normal text area then, and whenever we build a more feature rich editor we can still change it out

Comment (TimVanOnckelen) [reply web/src/lib/components/shared-components/settings/setting-textarea-wysiwyg.svelte:1]: Fair enough, i'll just use the text area

**Comment by TimVanOnckelen** [comment]
Changes stated in the reviews are added.

File: server/src/interfaces/notification.interface.ts [hunk 11]
```diff
@@ -70,18 +71,22 @@ export type EmailRenderRequest =
   | {
       template: EmailTemplate.TEST_EMAIL;
       data: TestEmailProps;
+      customTemplate?: string;
```
Comment (alextran1502) [inline server/src/interfaces/notification.interface.ts:74]: Should this be optional for all template?

Comment (TimVanOnckelen) [reply server/src/interfaces/notification.interface.ts:74]: I think it would be good to provide this for all templates already, even if they don't have an ui to override it. That way all templates behave the same.

Comment (alextran1502) [reply server/src/interfaces/notification.interface.ts:74]: Do you mean that this property will still be used for other email types, even without a template?

File: web/src/lib/components/admin-page/settings/template-settings/template-settings.svelte [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+    defaultConfig: SystemConfigDto;
+    config: SystemConfigDto;
+    disabled?: boolean;
+    onReset: SettingsResetEvent;
+    onSave: SettingsSaveEvent;
+  }
+
+  let { savedConfig, defaultConfig, config = $bindable(), disabled = false, onReset, onSave }: Props = $props();
+
+  let _htmlPreview = $state('');
```
Comment (alextran1502) [inline web/src/lib/components/admin-page/settings/template-settings/template-settings.svelte:32]: I think we don't need `_` to denote local state in Svelte

File: web/src/routes/admin/system-settings/+page.svelte [hunk 13]
```diff
@@ -180,6 +182,13 @@
       key: 'notifications',
       icon: mdiBellOutline,
     },
+    {
+      component: TemplateSettings,
```
Comment (alextran1502) [inline web/src/routes/admin/system-settings/+page.svelte:186]: I wonder if the template settings is within the email settings would be more logical?

File: server/src/interfaces/notification.interface.ts (hunk 11 above)
Comment (TimVanOnckelen) [reply server/src/interfaces/notification.interface.ts:74]: Yes indeed

File: web/src/routes/admin/system-settings/+page.svelte (hunk 13 above)
Comment (TimVanOnckelen) [reply web/src/routes/admin/system-settings/+page.svelte:186]: Will there be other notification types in the future apart from emails? Then it seems logical to me to split it up, so templates are generic over any notifcation type?

Comment (bo0tzz) [reply web/src/routes/admin/system-settings/+page.svelte:186]: Other notification types might not support a full templated body like emails do. For now I think it makes sense to be under emails, in the future we can always move it if necessary.

Comment (TimVanOnckelen) [reply web/src/routes/admin/system-settings/+page.svelte]: Ok, i'll change it :)

File: server/src/interfaces/notification.interface.ts (hunk 11 above)
Comment (alextran1502) [reply server/src/interfaces/notification.interface.ts:74]: Then, should the property here be specified as required instead of optional?

File: server/src/services/notification.service.ts [hunk 14]
```diff
@@ -175,14 +174,77 @@ export class NotificationService extends BaseService {
     return { messageId };
   }
 
+  async getTemplate(name: EmailTemplate, tempTemplate: string) {
```
Comment (alextran1502) [inline server/src/services/notification.service.ts:177]: Should `tempTemplate` here be replaced as `customTemplate`?

**Review by alextran1502 — APPROVED** [review]
